### ISSUE TITLE
Change pino settings after init

### ISSUE THREAD

**Issue by watson** [pinojs/pino#159]
I'd like to use pino in a module. It would be very convenient if I can just require a logger from anywhere in my module without having to pass around an already initialized logger.

But I also want the users of my module to be able to configure my logger, so in my main constructor I want to allow them to pass in options like the log-level and and output stream.

I found that I can overwrite the log-level using the `.level` property. But I can't overwrite the stream as far as I can tell. If this isn't possible, consider this a feature request 😃

Below is some simplified code showing what I'd like to achieve:

### `index.js`

```js
var logger = require('./lib/logger')
var foo = require('./lib/foo')

module.exports = MyModule

function MyModule (opts) {
  logger.level = opts.logLevel || 'info'
  logger.stream = opts.logFile || process.stdout
  foo.bar()
}
```

### `lib/logger.js`

```js
module.exports = require('pino')()
```

### `lib/foo.js`

```js
var logger = require('./logger')

exports.bar = function () {
  logger.info('logging all the things!')
}
```

**Comment by watson** [comment]
Also, of there's a better way to structure my code to achieve the same goal that I missed, please tell me 😃

**Comment by jsumners** [comment]
This sort of scenario is what https://www.npmjs.com/package/abstract-logging exists. Your module doesn't have to worry about what logger is supplied. The supplied logger just has to conform to the specified interface; which Pino does ;)

**Comment by watson** [comment]
@jsumners as far as I can see that would still require me to pass around a logger to all parts of my code that needs it which is what I'm trying to avoid. But I guess it would be a good idea not to make any assumptions about the logger and also allow the user to pass in his own logger. In that case I can't depend on a proprietary pino api to overwrite the output stream.

I think I'll re-write the `lib/logger.js` function to expose a property with the current instance of the logger, e.g:

```js
var pino = require('pino')
exports.i = pino() // `i` for instance
exports.init = function (opts, stream) {
  exports.i = pino(opts, stream)
}
```

At any time I can overwrite the `i` property with a user-supplied logger or I can call the `init` function with user-supplied options to configure the built-in logger.

Then I just have to do something like this every place I need to log:

```js
var log = require('./logger')

log.i.info('foobar')
```

Not pretty, but doable.

**Comment by jsumners** [comment]
It's possible that I don't fully understand the scenario. So I'll write what I'm thinking.

If I were writing a complex module that is meant to be used as a library, and I wanted to support logging, I'd do something like the following:

**index.js**
```js
var log
module.exports = function moduleInit (options) {
  log = require('./lib/logger')(options.logger)
}
```

**lib/logger.js**
```js
var logger
module.exports = function (instance) {
  if (logger) return logger
  logger = instance || require('abstract-logging')
}
```

**some-other-file.js**
```js
const log = require('./logger')()
log.trace('foo')
```

Given that `index.js` is the entry point, the `log` in `some-other-file.js` would have been initialized to either the supplied instance or the noop logger.

Is this the scenario you're trying to work with?

**Comment by watson** [comment]
I think your logger.js function is missing a few things, but I think I get the gist of it. I imagine you intended it to look something like this?

```diff
 var logger
 module.exports = function (instance) {
-  if (logger) return logger
-  logger = instance || require('abstract-logging')
+  if (!instance && logger) return logger
+  return logger = instance || require('abstract-logging')
 }
```

In that case, if `some-other-file.js` is required prior to running the `moduleInit` function, then its `log` variable will be set to the noop logger. Even after `moduleInit` is called later and sets up a proper logger later.

**Comment by jsumners** [comment]
My version assumes only one initialization, and it must occur via `require('the-module')`. But it is just an example to make sure I understand your use case. Would that sort of usage fix your problem? Or do you still think Pino needs more post-initialization customization features?

I have a strong suspicion that altering the stream after initialization is not a trivial matter. In particular #92 has discussion that would be directly affected.

**Comment by mcollina** [comment]
@watson I think you might want a pino-global thing. You should check out [bole](https://github.com/rvagg/bole/blob/master/bole.js#L2) and https://www.npmjs.com/package/individual. I think it can be a nice module on its own, rather than adding features here.

Altering the stream after initialization is *bad* if extreme mode is used. However it should be fine if that is turned off.

**Comment by watson** [comment]
@jsumners No, unfortunately I need to be able to configure pino using user supplied options _after_ all the files have been required and already contain a reference to the logger. But I can get around it using the code written in [my previous comment](https://github.com/pinojs/pino/issues/159#issuecomment-270123325), so it's no biggie.

And if changing the stream mid-way is problematic anyway, I can understand why this can't be supported. Plus I imagine my specific use-case it pretty unique, so feel free to close the issue 😃

@mcollina 👍
